vpr/src/route/rr_graph_fpga_interchange.cpp

```diff
@@ -0,0 +1,1179 @@
+                    node_* temp = nullptr;
+                    if (vertex->links[i] != nullptr) {
+                        if (i == 2 || i == 3)
+                            cnt--;
+                        temp = vertex->links[i];
+                        temp->links[(i + 2) % 4] = nullptr;
+                    }
+                }
+                delete (vertex);
+            } else if (vertex->on_route) {
+                vertex->visited->on_route = true;
+                if (vertex->visited->segment_id == -1)
+                    vertex->visited->segment_id = vertex->segment_id;
+            }
+        }
+        return cnt;
+    }
+
+    void populate_chan_loc_map(int node_id, node_* root, float R, float C) {
+        /*
+         * Create future rr_nodes that corespond to given FPGA Interchange node
+         */
+        std::map<std::tuple<location, e_rr_type /*CHANX/CHANY*/>, int /*index*/> loc_chan_idx;
+
+        std::queue<node_*> walker;
+        std::stack<node_*> back_walker;
+        walker.emplace(root);
+        bool all_nulls, is_chanx, is_chany;
+        while (!walker.empty()) {
+            node_* vertex = walker.front();
+            walker.pop();
+            back_walker.emplace(vertex);
+            is_chany = false;
+            is_chanx = vertex->has_pins;
+            all_nulls = true;
+            for (int i = 0; i < 4; ++i) {
+                if (vertex->links[i] != nullptr) {
+                    all_nulls = false;
+                    if (vertex->links[i]->visited == vertex || vertex->links[i] == vertex->visited) {
+                        is_chanx = i == 0 ? true : is_chanx;
+                        is_chany = i == 1 ? true : is_chany;
+                    }
+                    if (vertex->links[i]->visited == vertex) {
+                        walker.emplace(vertex->links[i]);
+                    }
+                }
+            }
+            is_chanx = all_nulls ? all_nulls : is_chanx;
+            std::tuple<location, e_rr_type> key1(vertex->loc, e_rr_type::CHANY);
+            std::tuple<location, e_rr_type> key2(location(std::get<0>(vertex->loc), std::get<1>(vertex->loc) - 1), e_rr_type::CHANX);
+
+            if (is_chany) {
+                loc_chan_idx[key1] = chan_loc_map_[key1].size();
+                chan_loc_map_[key1].emplace_back(vertex->segment_id, R, C);
+            }
+            if (is_chanx) {
+                loc_chan_idx[key2] = chan_loc_map_[key2].size();
+                chan_loc_map_[key2].emplace_back(vertex->segment_id, R, C);
+            }
+            if (is_chanx && is_chany) {
+                shorts_.emplace_back(std::get<0>(key1), std::get<1>(key1), loc_chan_idx[key1],
+                                     std::get<0>(key2), std::get<1>(key2), loc_chan_idx[key2]);
+            }
+        }
+
+        /*
+         * Create edges between future rr_nodes that correspond to the given FPGA Interchange node
+         */
+        while (!back_walker.empty()) {
```

Comment:
In the body of this while, replace all `links[<interger index>]` with `links[<enum value>]`